File: Data Structures & Algorithms/same-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
    isSameTree(p, q) {
        const f = (a, b) => {
            if (a == null && b == null) return true
            if (a == null || b == null) return false
            if (a.val == b.val) return f(a.left, b.left) && f(a.right, b.right)
            return false
        }

        return f(p, q)
    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Codec:
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        lst = []

        def dfs(node: Optional[TreeNode]):
            if node is None:
                lst.append("null")
                return

            lst.append(str(node.val))

            dfs(node.left)
            dfs(node.right)

        dfs(root)

        return ",".join(lst)

    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        lst = data.split(",")
        print(lst)

        if len(list(filter(lambda x: x != "null", lst))) == 0:
            return None

        root = TreeNode(int(lst[0]))
        stack = [[root, 2]]

        for i in range(1, len(lst)):
            parent = stack[-1]
            parent[1] -= 1

            if parent[1] == 0:
                stack.pop()

            node = lst[i]
            if node != "null":
                tree = TreeNode(int(node))
                if parent[1] == 1:
                    parent[0].left = tree
                else:
                    parent[0].right = tree
                stack.append([tree, 2])

        return root
